### DISCUSSION TITLE
Should properties be exposed as attributes by default?

### DISCUSSION THREAD

**Discussion by RandomEngy** [lit/lit#2497] [General]
I've generally steered clear of attributes for the most part:
* Type conversion is a pain, impossible to pass functions
* It's easier to have all inputs come in the same way as properties
* It's more overhead to expose as an attribute
* It's easy to set properties in parent templates with the `.propName=` syntax

I'm coming from using [FAST Element](https://www.fast.design/docs/fast-element/getting-started): the way it works there is you can expose a property with `@observable` or in addition expose it as an attribute with `@attr`.

So I was a bit surprised to find that Lit exposes all properties as attributes by default. I'm going to need to litter my code with `@property({ attribute: false })` all over the place. I can also imagine problems with component authors unwittingly exposing attributes with odd or unhelpful squashed-down names like `somepropertyhere`, then having issues with backwards-compatibility if they want to fix them.

I wanted to see how others approach things: do people make heavy use of the attributes?

**Comment by zandaqo** [comment]
> I wanted to see how others approach things: do people make heavy use of the attributes?

tldr: I use quite a few properties (as in class fields) in my components, but few of them are Lit's *reactive properties* (like `@property`). And the reactive ones do tend to be attributes represented by booleans, strings, and numbers (in that order of frequency).

Though, that's because I separate my domain (business logic) from the UI state and use reactive properties for UI state only, while using my own proxy based state management for the domain objects ([compago](https://github.com/zandaqo/compago#advanced-state-management-with-observable)). For example, if I'm building a blog, I have an object (aka entity) called Article and it can be complex with nested objects, arrays, etc. On the client side, the object is wrapped into a proxy (`Observable`) that triggers events whenever and however the object's contents change. The web component listens to the observable and re-renders. Meanwhile, the Lit reactive properties only hold the UI state, like `isEditing`, `isLoaded`, and the components inputs/API, for example `articleId`--I might set the article id in the component's parent so that fetching the actual article is handled by the component itself.

**Comment by UserGalileo** [comment]
For UI components (button, accordion, tabs, card...) I think most properties tend to be reflected to attributes cause they're usually primitives, so it's convenient. When I'm using Lit for an entire application though, I agree, I use mainly properties and I don't care at all about attributes, cause I usually pass objects, arrays, functions, etc...

**Comment by justinfagnani** [comment]
Not that Lit creates observed attributes by default, but does not _reflect_ properties to attributes by default.

We do this because so many reactive properties are simple types that we can deserialize to properties and letting them be set by attributes has no cost and makes the components more usable to plain HTML consumers. We don't reflect by default because there is a performance cost to it and ideally attributes would be "owned" and set only by their container, not the component itself. Reflected attributes are mostly needed for styling, so we want people to choose which properties should be reflected carefully.

We also have `@state()` which is intended for private state and doesn't observe attributes.

So regarding just the default to observe attributes:

I'd say that the first point is true: for complex-valued objects there isn't a great default to deserialize, so I do end up with _some_ `{attribute: false}` in my components, but I definitely don't add that for simple types.

The easier part I don't think is true because from the component POV it _is_ the property being set - Lit takes care of that.

The performance part isn't true because there's already a generic attributeToProperty function - marking a property as settable as an attribute doesn't add any code or compute cost when it's not used.
